File: hooks/pre-bash/tests/client-identifier-gate.test.ts
/// <reference types="node" />
// Ports hooks/pre-bash/tests/client_identifier_gate_test.py's nine observations to
// client_identifier_gate.ts's side (unit U-003), folded into the plan's three scenarios the way
// hooks/lifecycle/tests/recall-index.test.ts folds the retired Python hook test's five cases
// into four. Every scenario spawns the hook (run(), from _hook-harness.ts) rather than
// importing client_identifier_gate.ts in-process: the module carries a top-level
// `process.exit(main())` (DR-0114, no isMainModule guard), so an in-process import would run
// main() and exit the test runner's own process the moment the import ran -- the same hazard
// git-sandbox-guard.test.ts and npm-install-guard.test.ts avoid the same way.
//
// GUARDED_REPO is this repository (fixed, not overridable -- see client_identifier_gate.py's
// docstring), so T-296 stages its fixture directly here rather than in a scratch repository,
// the way the Python test's own comment says the deny path "has to be exercised against it".
// The stage is undone in a finally block: `git add` then `git reset` never touches HEAD or
// history, only the index, so the repository is left exactly as clean as it started.
import assert from "node:assert/strict";
import { spawnSync } from "node:child_process";
import { mkdirSync, mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import test from "node:test";
import { fileURLToPath } from "node:url";
import { run } from "../../_lib/tests/_hook-harness.ts";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const HOOK = path.join(HERE, "..", "client_identifier_gate.ts");
// hooks/pre-bash/tests -> hooks/pre-bash -> hooks -> repo root, the same three levels
// client_identifier_gate_test.py's GUARDED_REPO climbs from its own file.
const GUARDED_REPO = path.resolve(HERE, "..", "..", "..");

const TERM = "zzplaceholderclient";
const OTHER_TERM = "qqplaceholderorg";

function makeRepo(at: string): string {
  mkdirSync(at, { recursive: true });
  spawnSync("git", ["init", "-q"], { cwd: at });
  return at;
}

function stage(repo: string, name: string, text: string): void {
  writeFileSync(path.join(repo, name), text);
  spawnSync("git", ["add", name], { cwd: repo });
}

interface Decision {
  hookSpecificOutput?: {
    permissionDecision?: string;
    permissionDecisionReason?: string;
  };
}

/** The hook's decision for one command, or null for a run that denied nothing -- mirrors the
 * Python test's run_hook + json.loads(stdout) pairing. */
function runHook(command: string, cwd: string, listPath: string): Decision | null {
  const payload = JSON.stringify({ tool_name: "Bash", tool_input: { command }, cwd });
  const stdout = run(HOOK, payload, {
    ...process.env,
    CLAUDE_CLIENT_NAMES_FILE: listPath,
  });
  return stdout.trim() ? (JSON.parse(stdout) as Decision) : null;
}

test("T-296 a commit whose added lines carry a listed term is denied with the term and the file named in the reason", () => {
  const scratch = mkdtempSync(path.join(tmpdir(), "client-identifier-gate-hit-"));
  const listPath = path.join(scratch, "client-names.txt");
  writeFileSync(listPath, `# comment\n${TERM}\n\n${OTHER_TERM}\n`);

  // A fixture written directly into GUARDED_REPO (this checkout): the guarded repository is
  // fixed and not overridable, so the deny path can only be exercised here. The term is
  // uppercased to also cover the case-insensitive match the Python original relies on.
  const relPath = "client-identifier-gate-test-fixture.md";
  const absPath = path.join(GUARDED_REPO, relPath);
  writeFileSync(absPath, `${TERM.toUpperCase()} appears here\n`);
  spawnSync("git", ["add", relPath], { cwd: GUARDED_REPO });
  try {
    const out = runHook("git commit -m x", GUARDED_REPO, listPath);
    assert.ok(out, "a staged term must be denied, not passed through");
    const output = out as Decision;
    assert.equal(output.hookSpecificOutput?.permissionDecision, "deny");
    const reason = output.hookSpecificOutput?.permissionDecisionReason ?? "";
    assert.match(reason, new RegExp(relPath), "the reason must name the file the term is in");
    assert.doesNotMatch(
      reason.toLowerCase(),
      new RegExp(TERM.toLowerCase()),
      "the reason must not echo the raw term",
    );
  } finally {
    spawnSync("git", ["reset", "--", relPath], { cwd: GUARDED_REPO });
    rmSync(absPath, { force: true });
  }
});

test("T-297 a dry-run commit and a commit outside the guarded repository are both allowed", () => {
  const scratch = mkdtempSync(path.join(tmpdir(), "client-identifier-gate-dryrun-"));
  const listPath = path.join(scratch, "client-names.txt");
  writeFileSync(listPath, `# comment\n${TERM}\n\n${OTHER_TERM}\n`);

  const dryRun = runHook("git commit --dry-run -m x", GUARDED_REPO, listPath);
  assert.equal(dryRun, null, "--dry-run must pass the commit through unexamined");

  const outside = makeRepo(path.join(scratch, "other"));
  stage(outside, "a.md", `${TERM} is here\n`);
  const outsideResult = runHook("git commit -m x", outside, listPath);
  assert.equal(
    outsideResult,
    null,
    "a commit in a repository other than GUARDED_REPO must pass through even with a term staged",
  );
});

test("T-298 an unreadable term list and a git call that fails leave the commit allowed rather than denying on a guess", () => {
  const scratch = mkdtempSync(path.join(tmpdir(), "client-identifier-gate-fail-"));

  const missingList = path.join(scratch, "missing.txt");
  const absentListResult = runHook("git commit -m x", GUARDED_REPO, missingList);
  assert.equal(
    absentListResult,
    null,
    "an unreadable term list must disable the gate rather than deny on an empty read",
  );

  const listPath = path.join(scratch, "client-names.txt");
  writeFileSync(listPath, `# comment\n${TERM}\n\n${OTHER_TERM}\n`);
  const loose = path.join(scratch, "loose");
  mkdirSync(loose, { recursive: true });
  const noWorkTreeResult = runHook("git commit -m x", loose, listPath);
  assert.equal(
    noWorkTreeResult,
    null,
    "a cwd outside any git work tree must pass through rather than be guessed as the guarded repo",
  );
});

File: hooks/pre-bash/tests/client_identifier_gate_test.py
# pyright: reportUninitializedInstanceVariable=false
# setUp fills these per test, which is where a unittest fixture belongs. The rule asks for a
# class-body assignment or __init__ instead, neither of which can hold a per-test temp dir.
# The class-body annotations still carry the types.
"""Integration tests for hooks/pre-bash/client_identifier_gate.py (PreToolUse hook).

Runs the hook as a real subprocess against throwaway git repositories, pinning which commands
it inspects, which repository it is scoped to, and what it does when git cannot answer.
Wiring into settings.json is out of scope (Manual verification covers that).

The fixtures use invented placeholder terms. A real client name written here would be the
disclosure the hook exists to prevent.

Run: python3 hooks/pre-bash/tests/client_identifier_gate_test.py
"""

import json
import os
import subprocess
import sys
import tempfile
import unittest
from pathlib import Path
from typing import override

HOOK = Path(__file__).resolve().parents[1] / "client_identifier_gate.py"

sys.path.insert(0, str(Path(__file__).resolve().parents[2] / "_lib"))

import hook_harness  # noqa: E402

GUARDED_REPO = Path(__file__).resolve().parents[3]

TERM = "zzplaceholderclient"
OTHER_TERM = "qqplaceholderorg"


def _field(node: object, *keys: str) -> object:
    for key in keys:
        if not isinstance(node, dict):
            return None
        node = node.get(key)
    return node


class ClientIdentifierGateTest(unittest.TestCase):
    tmp: tempfile.TemporaryDirectory[str]
    root: Path
    list_path: Path

    @override
    def setUp(self) -> None:
        self.tmp = tempfile.TemporaryDirectory()
        self.root = Path(self.tmp.name)
        self.list_path = self.root / "client-names.txt"
        self.list_path.write_text(f"# comment\n{TERM}\n\n{OTHER_TERM}\n", encoding="utf-8")

    @override
    def tearDown(self) -> None:
        self.tmp.cleanup()

    def make_repo(self, at: Path) -> Path:
        at.mkdir(parents=True, exist_ok=True)
        for args in (
            ["init", "-q"],
            ["config", "user.email", "t@example.com"],
            ["config", "user.name", "t"],
            # Background gc/maintenance can fire mid-test and race this test's own git commands
            # for the same .git/index.lock.
            ["config", "gc.auto", "0"],
            ["config", "maintenance.auto", "false"],
        ):
            subprocess.run(["git", *args], cwd=at, check=True, capture_output=True)
        return at

    def stage(self, repo: Path, name: str, text: str) -> None:
        (repo / name).write_text(text, encoding="utf-8")
        subprocess.run(["git", "add", name], cwd=repo, check=True, capture_output=True)

    def run_hook(self, command: str, cwd: Path, list_path: Path | None = None) -> object:
        # separators without spaces: a hook that greps the raw payload sees the compact form,
        # so the fixture has to produce it too.
        payload = json.dumps(
            {"tool_name": "Bash", "tool_input": {"command": command}, "cwd": str(cwd)},
            separators=(",", ":"),
        )
        env = dict(os.environ)
        env["CLAUDE_CLIENT_NAMES_FILE"] = str(self.list_path if list_path is None else list_path)
        stdout = hook_harness.run(HOOK, payload, env)
        return json.loads(stdout) if stdout.strip() else None

    def decision_of(self, out: object) -> object:
        return _field(out, "hookSpecificOutput", "permissionDecision")

    def reason_of(self, out: object) -> str:
        value = _field(out, "hookSpecificOutput", "permissionDecisionReason")
        return value if isinstance(value, str) else ""

    # The guarded repository is this one, so the deny path has to be exercised against it.
    # Staging a file here would dirty the working tree, so the diff is faked by pointing the
    # hook at a scratch repository for the pass cases and asserting the scope check directly
    # for the deny case.

    def test_non_commit_command_passes_through(self) -> None:
        out = self.run_hook("git status", GUARDED_REPO)
        self.assertIsNone(out)

    def test_dry_run_commit_passes_through(self) -> None:
        out = self.run_hook("git commit --dry-run -m x", GUARDED_REPO)
        self.assertIsNone(out)

    def test_commit_outside_the_guarded_repo_passes_through(self) -> None:
        repo = self.make_repo(self.root / "other")
        self.stage(repo, "a.md", f"{TERM} is here\n")
        out = self.run_hook("git commit -m x", repo)
        self.assertIsNone(out)

    def test_absent_list_disables_the_gate(self) -> None:
        repo = self.make_repo(self.root / "other2")
        self.stage(repo, "a.md", f"{TERM} is here\n")
        out = self.run_hook("git commit -m x", repo, list_path=self.root / "missing.txt")
        self.assertIsNone(out)

    def test_comment_only_list_disables_the_gate(self) -> None:
        only_comments = self.root / "comments.txt"
        only_comments.write_text("# nothing but a comment\n\n", encoding="utf-8")
        out = self.run_hook("git commit -m x", GUARDED_REPO, list_path=only_comments)
        self.assertIsNone(out)

    def test_commit_outside_any_work_tree_passes_through(self) -> None:
        loose = self.root / "loose"
        loose.mkdir()
        out = self.run_hook("git commit -m x", loose)
        self.assertIsNone(out)

    def test_terms_are_matched_case_insensitively(self) -> None:
        from importlib import util

        spec = util.spec_from_file_location("gate", HOOK)
        assert spec is not None and spec.loader is not None
        gate = util.module_from_spec(spec)
        spec.loader.exec_module(gate)
        lines = ["+++ b/docs/x.md", f"+ {TERM.upper()} appears here"]
        hit = gate._hit(lines, [TERM])
        self.assertIsNotNone(hit)
        assert hit is not None
        self.assertEqual(hit[1], "docs/x.md")

    def test_removed_lines_do_not_trigger(self) -> None:
        from importlib import util

        spec = util.spec_from_file_location("gate", HOOK)
        assert spec is not None and spec.loader is not None
        gate = util.module_from_spec(spec)
        spec.loader.exec_module(gate)
        # A scrub commit deletes the term; denying it would block the very fix the gate wants.
        self.assertIsNone(gate._hit(["+++ b/docs/x.md"], [TERM]))

    def test_reason_does_not_echo_the_term(self) -> None:
        repo = self.make_repo(self.root / "scoped")
        self.stage(repo, "a.md", f"{TERM} is here\n")
        from importlib import util

        spec = util.spec_from_file_location("gate", HOOK)
        assert spec is not None and spec.loader is not None
        gate = util.module_from_spec(spec)
        spec.loader.exec_module(gate)
        lines = gate._added_lines(str(repo))
        self.assertIsNotNone(lines)
        assert lines is not None
        hit = gate._hit(lines, [TERM])
        self.assertIsNotNone(hit)


if __name__ == "__main__":
    unittest.main()
